### PR TITLE
[FIX] hr_attendance:  Calculating extra hours based on timezone

### PR DESCRIPTION
**Pull request by bona-odoo**
**Steps to reproduce:**
- open attendance module
- go to overview and create a new attendance within 5:30 AM

**Issue:**
- Extra hours[overtime_hours] are not calculated accurately[output = 0].

**Cause:**
The [SQL Query](https://github.com/odoo/odoo/blob/17.0/addons/hr_attendance/models/hr_attendance.py#L77-L91)
[condition : date_trunc('day',att.check_in) = date_trunc('day', ot.date)] compares `att.check_in` time stored in UTC with `ot.date` stored in employee timezone, leading to incorrect calculations.

**Solution:**
Modify SQL query to convert `att.check_in` time to employee timezone before comparison with `ot.date`.

In saas-16.4, This [SQL Query](https://github.com/odoo/odoo/blob/saas-16.4/addons/hr_attendance/report/hr_attendance_report.py#L40-L46) is responsible for converting `check_in` time to employee 
timezone.

Description of the issue/feature this PR addresses:

**Current behavior before PR:**

RUNBOT-17.0
runbot_17

**Desired behavior after PR is merged:**

image

UPG - 1573228
TASK - 3900287

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_attendance/models/hr_attendance.py [hunk 1]
```diff
@@ -83,8 +83,13 @@ def _compute_overtime_hours(self):
...
              INNER JOIN hr_attendance_overtime ot
-                    ON date_trunc('day',att.check_in) = date_trunc('day', ot.date)
-                    AND date_trunc('day',att.check_out) = date_trunc('day', ot.date)
+                    ON date_trunc('day',CAST(att.check_in
+                                                at time zone 'utc'
+                                                at time zone
+                                                (SELECT calendar.tz FROM resource_calendar calendar
+                                                 INNER JOIN hr_employee employee ON employee.id = att.employee_id
+                                                 WHERE calendar.id = employee.resource_calendar_id)
+                                        as date)) = date_trunc('day', ot.date)
```
Comment (jugj-odoo) [inline addons/hr_attendance/models/hr_attendance.py:92]: For efficiency and readability i think we should maybe pre-compute this table with something like this
```sql
WITH employee_timezones AS (
    SELECT
        employee.id AS employee_id,
        calendar.tz AS timezone
    FROM
        hr_employee employee
    INNER JOIN
        resource_calendar calendar ON employee.resource_calendar_id = calendar.id
    WHERE
        employee.id IN %s
)
SELECT
    att.id AS att_id,
    att.worked_hours AS att_wh,
    ot.id AS ot_id,
    ot.duration AS ot_d,
    ot.date AS od,
    att.check_in AS ad
FROM
    hr_attendance att
INNER JOIN
    hr_attendance_overtime ot ON att.employee_id = ot.employee_id
INNER JOIN
    employee_timezones etz ON att.employee_id = etz.employee_id
.....
```

**Review by jugj-odoo — CHANGES_REQUESTED** [review]
Hello thanks for handling this painful bug, we love time zones :D , small remark about the SQL query, and could you also add a test to ensure that this is not broken in the future ? Thanks :D

File: addons/hr_attendance/models/hr_attendance.py (hunk 1 above)
Comment (bona-odoo) [reply addons/hr_attendance/models/hr_attendance.py]: I have made the suggested changes to the SQL query and added the test case.

Comment (jugj-odoo) [reply addons/hr_attendance/models/hr_attendance.py]: @bona-odoo looks good, we'll r+ this after the freeze to not conflict a current attendance task

**Comment by sofiagvaladze** [comment]
robodoo r+

**Pull request closed without merging**